### DISCUSSION TITLE
[Contentful] GraphQL Typegen is marking required fields as nullable

### DISCUSSION THREAD

**Discussion by mwskwong** [gatsbyjs/gatsby#36292] [Help]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

I'm not too sure whether this is considered a bug or how feasible for it to be "fixed".

In Contentful, it is possible to mark certain fields "required", as the name suggests, users must fill in a value of such fields before publishing the content. However, in Gatsby, the generated type for the GraphQL query appears to be ignored/unable to pick up this property. All fields of the query will always have the type of  `<string | number | object | etc...> | null` regardlessly.

### Reproduction Link

N/A

### Steps to Reproduce

1. Create a gatsby project with Contentful
2. Create a content model with one or many required fields in Contentful
3. Query the content in Gatsby

### Expected Result

"Required" filed is not nullable

### Actual Result

"Required" filed is nullable

### Environment

```shell
System:
    OS: Windows 10 10.0.19044
    CPU: (8) x64 Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
  Binaries:
    Node: 17.9.1 - C:\Program Files\nodejs\node.EXE
    npm: 8.11.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (101.0.1210.53)
  npmPackages:
    gatsby: ^4.19.2 => 4.19.2 
    gatsby-plugin-emotion: ^7.19.0 => 7.19.0 
    gatsby-plugin-image: ^2.19.0 => 2.19.0 
    gatsby-plugin-manifest: ^4.19.0 => 4.19.0 
    gatsby-plugin-material-ui: ^4.1.0 => 4.1.0 
    gatsby-plugin-netlify: ^5.0.1 => 5.0.1 
    gatsby-plugin-offline: ^5.19.0 => 5.19.0 
    gatsby-plugin-perf-budgets: ^0.0.18 => 0.0.18 
    gatsby-plugin-robots-txt: ^1.7.1 => 1.7.1 
    gatsby-plugin-sharp: ^4.19.0 => 4.19.0 
    gatsby-plugin-sitemap: ^5.19.0 => 5.19.0 
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.27 => 1.1.27 
    gatsby-source-contentful: ^7.17.0 => 7.17.0 
    gatsby-transformer-sharp: ^4.19.0 => 4.19.0
```

### Config Flags

FAST_DEV: true,
PARALLEL_SOURCING: true,
PRESERVE_FILE_DOWNLOAD_CACHE: true

**Comment by LekoArts** [comment]
You have to also type the GraphQL schema this way. That's the source of truth, not the Contentful side of things.

**Reply by mwskwong** [reply]
I don't follow, the types are supposed to be generated programmatically since `graphqlTypegen` is enabled. How do you "type the GraphQL schema"?

**Reply by LekoArts** [reply]
https://www.gatsbyjs.com/docs/how-to/local-development/graphql-typegen/#non-nullable-types

**Reply by mwskwong** [reply]
I'm really new to this. If I'm not mistaken, let's say I have a Contentful content as follows
```graphql
contentfulName {
    firstName
    middleName
    lastName
}
```

If `firstName` and `lastName` are non-nullable, then I should do
```ts
export const createSchemaCustomization: GatsbyNode["createSchemaCustomization"] = ({ actions }) => {
  actions.createTypes(`
    type ContentfulName {
      firstName: String!
      lastName: String!
    }
  `);
};
```

Is that the case?

**Reply by LekoArts** [reply]
Yeah, correct. You can use GraphiQL (`localhost:8000/___graphql`) to find out the `ContentfulName` type name. Either by querying `__typename` or looking at the right "Docs" sidebar

![image](https://user-images.githubusercontent.com/16143594/182084654-698e4f9b-9136-4ebb-8720-48e4069eafba.png)

And the type has to match, if your `firstName` is a string, then yeah, it should be `String!`. If it's e.g. a number, it's `Int!` etc.

**Reply by mwskwong** [reply]
Faced some weird issue, I define one of the content types as follow, knowing all the fields are "required"
```graphql
    type ContentfulPdf {
      title: String!
      file: ContentfulAsset!
      thumbnail: ContentfulAsset!
    }
```

Gatsby somehow complains `Cannot return null for non-nullable field ContentfulPdf.file.`.
I query `allContentfulPdf` just to be sure, and no, there aren't any `null` values at all.

One thing I found in common is it only seems to affect "Long Text" and "Asset" data types.

I have no idea what Gatsby is complaining about.
